lib/phoenix_ecto/plugs/check_repo_ready.ex

```diff
@@ -0,0 +1,64 @@
+defmodule Phoenix.Ecto.Plugs.CheckRepoReady do
+  @moduledoc """
+  A plug that does some checks on your application repos.
+
+  Checks if the storage is up (database is created) or if there are any pending migrations.
+  Both checks can raise an error if the conditions are not met.
+
+  ## Plug options
+
+  * `:otp_app` - name of the application which the repos are fetched from
+  * `:get_migration_function` - function that returns the migrations for a given `repo`.
+    Should return a list of tuples with three elements,
+    e.g. `[{state :: :up | :down, version :: integer(), name :: String.t()]]`
+    If `ecto_sql` dependecy is loaded, uses `Ecto.Migrator.migrations/1` by default.
+  """
+
+  @behaviour Plug
+
+  alias Plug.Conn
+
+  def init(opts), do: opts
+
+  def call(%Conn{} = conn, opts) do
+    repos = Application.get_env(opts[:otp_app], :ecto_repos, [])
+
+    for repo <- repos, Process.whereis(repo) do
+      check_storage_up!(repo)
+      check_pending_migrations!(repo, opts)
+    end
+
+    conn
+  end
+
+  defp check_storage_up!(repo) do
+    adapter_implement_storage_status? =
+      Code.ensure_loaded?(repo.__adapter__) &&
+        function_exported?(repo.__adapter__, :storage_status, 1)
+
+    storage_status_down? = fn -> repo.__adapter__.storage_status(repo.config()) == :down end
+
+    if adapter_implement_storage_status? && storage_status_down?.(),
+      do: raise(Phoenix.Ecto.StorageNotCreatedError, repo: repo)
```

Comment:
I think it is important for this function to only raise this exception and nothing more. So I would rewrite it to something like this:

```elixir
  try do
    if Code.ensure_loaded?(repo.__adapter__) &&
         function_exported?(repo.__adapter__, :storage_status, 1) do
      repo.__adapter__.storage_status(repo.config())
    else
      {:error, :not_available}
    end
  rescue 
    _ -> :ok
  else
    :down -> raise Phoenix.Ecto.StorageNotCreatedError, repo: repo
    _ -> :ok
  end
```

Similar has to be done to the function below.

Reply:
Makes sense!

[fixed here](https://github.com/phoenixframework/phoenix_ecto/compare/5083d99f7f515143f5d00420d620f5491833597f..693fbe89c367841069104dcf2070e1e40f9f0028)